tests/test_read_lazy.py

```diff
@@ -0,0 +1,592 @@
+    read_lazy(store)
+    # 4 is number of chunks
+    store.assert_access_count("obs/_index", 4)
+
+
+def test_access_count_dtype(
+    adata_remote_with_store_tall_skinny: tuple[AnnData, AccessTrackingStore],
+):
+    remote, store = adata_remote_with_store_tall_skinny
+    store.initialize_key_trackers(["obs/cat/categories"])
+    store.assert_access_count("obs/cat/categories", 0)
+    # This should only cause categories to be read in once
+    remote.obs["cat"].dtype
+    remote.obs["cat"].dtype
+    remote.obs["cat"].dtype
+    store.assert_access_count("obs/cat/categories", 1)
+
+
+def test_uns_uses_dask(adata_remote_orig: tuple[AnnData, AnnData]):
+    remote, _ = adata_remote_orig
+    assert isinstance(remote.uns["nested"]["nested_further"]["array"], DaskArray)
+
+
+def test_to_memory(adata_remote_orig: tuple[AnnData, AnnData]):
+    remote, orig = adata_remote_orig
+    remote_to_memory = remote.to_memory()
+    assert_equal(remote_to_memory, orig)
+
+
+def test_view_to_memory(adata_remote_orig: tuple[AnnData, AnnData]):
+    remote, orig = adata_remote_orig
+    subset_obs = orig.obs["obs_cat"] == "a"
+    assert_equal(orig[subset_obs, :], remote[subset_obs, :].to_memory())
+
+    subset_var = orig.var["var_cat"] == "a"
+    assert_equal(orig[:, subset_var], remote[:, subset_var].to_memory())
+
+
+def test_view_of_view_to_memory(adata_remote_orig: tuple[AnnData, AnnData]):
+    remote, orig = adata_remote_orig
+    subset_obs = (orig.obs["obs_cat"] == "a") | (orig.obs["obs_cat"] == "b")
+    subsetted_adata = orig[subset_obs, :]
+    subset_subset_obs = subsetted_adata.obs["obs_cat"] == "b"
+    subsetted_subsetted_adata = subsetted_adata[subset_subset_obs, :]
+    assert_equal(
+        subsetted_subsetted_adata,
+        remote[subset_obs, :][subset_subset_obs, :].to_memory(),
+    )
+
+    subset_var = (orig.var["var_cat"] == "a") | (orig.var["var_cat"] == "b")
+    subsetted_adata = orig[:, subset_var]
+    subset_subset_var = subsetted_adata.var["var_cat"] == "b"
+    subsetted_subsetted_adata = subsetted_adata[:, subset_subset_var]
+    assert_equal(
+        subsetted_subsetted_adata,
+        remote[:, subset_var][:, subset_subset_var].to_memory(),
+    )
+
+
+def test_unconsolidated(tmp_path: Path, mtx_format):
+    adata = gen_adata((1000, 1000), mtx_format)
+    orig_pth = tmp_path / "orig.zarr"
+    adata.write_zarr(orig_pth)
+    (orig_pth / ".zmetadata").unlink()
+    store = AccessTrackingStore(orig_pth)
+    store.initialize_key_trackers(["obs/.zgroup", ".zgroup"])
+    with pytest.warns(UserWarning, match=r"Did not read zarr as consolidated"):
+        remote = read_lazy(store)
+    remote_to_memory = remote.to_memory()
+    assert_equal(remote_to_memory, adata)
+    store.assert_access_count("obs/.zgroup", 1)
+
+
+# remote has object dtype, need to convert back for integers booleans etc.
+def correct_extension_dtype_differences(remote: pd.DataFrame, memory: pd.DataFrame):
```

Comment:
why does remote have object dtypes? can that be fixed ahead of time? will that affect users or is that just here in the tests?

also for clarity, please call it `fix_extension_dtype_differences` or `unify_extension_dtypes` and make the comment a docstring. I was unsure if `correct` was meant as a verb or something like `make_correct_...`

Reply:
I will change the name and add a better comment since it's only used for concat.  There is no way of concat-ing these lazy pandas adapter arrays we have so in order to work with them, we present a dask array wrapping each when they are concatenated.  Practically, this actually doesn't change anything.  But it means that when we make the round trip, we lose the original data type.  We could maybe store this in `uns` or something (what the original data type was) but

1. People concating remote/large datasets probably won't be reading them into memory very often
2. This gets complicated quickly with mixed data types.  So if you have two similarly named columns but different numerical data types, we need to start dealing with upcasting/downcasting.

Feels like a cross-that-bridge-when-we-come-to-it issue